src/executables/compactdb_common.c

```diff
@@ -0,0 +1,271 @@
+	  goto error;
+	}
+
+      sm_downcase_name (class_names[i], downcase_class_name, SM_MAX_IDENTIFIER_LENGTH);
+
+      class_ = locator_find_class (downcase_class_name);
+      if (class_ != NULL)
+	{
+	  (*class_list)[(*num_class_list)] = class_;
+	  (*num_class_list)++;
+	}
+      else
+	{
+	  printf (msgcat_message (MSGCAT_CATALOG_UTILS, MSGCAT_UTIL_SET_COMPACTDB, COMPACTDB_MSG_CLASS),
+		  downcase_class_name);
+
+	  printf (msgcat_message (MSGCAT_CATALOG_UTILS, MSGCAT_UTIL_SET_COMPACTDB, COMPACTDB_MSG_INVALID_CLASS));
+	}
+    }
+
+  return NO_ERROR;
+
+error:
+  if (*class_list)
+    {
+      free (*class_list);
+      *class_list = NULL;
+    }
+
+  if (num_class_list)
+    {
+      *num_class_list = 0;
+    }
+
+  return ER_FAILED;
+}
+
+/*
+ * get_class_mops_from_file - Get a list of class mops from specified file
+ *    return: error code
+ *    input_filename(in): input file name
+ *    class_list(out): pointer to returned list of class mops
+ *    num_class_mops(out): pointer to returned number of mops
+ */
+int
+get_class_mops_from_file (const char *input_filename, MOP ** class_list, int *num_class_mops)
+{
+  int status = NO_ERROR;
+  int i = 0;
+  FILE *input_file;
+  char buffer[DB_MAX_IDENTIFIER_LENGTH];
+  char **class_names = NULL;
+  int num_class = 0;
+  int len = 0;
+  char *ptr = NULL;
+
+  if (input_filename == NULL || class_list == NULL || num_class_mops == NULL)
+    {
+      return ER_FAILED;
+    }
+
+  *class_list = NULL;
+  *num_class_mops = 0;
+  if (get_num_requested_class (input_filename, &num_class) != NO_ERROR)
+    {
+      return ER_FAILED;
+    }
+
+  if (num_class == 0)
+    {
+      return ER_FAILED;
+    }
+
+  input_file = fopen (input_filename, "r");
+  if (input_file == NULL)
+    {
+      perror (input_filename);
+      return ER_FAILED;
+    }
+
+  class_names = (char **) malloc (DB_SIZEOF (char *) * num_class);
+  if (class_names == NULL)
+    {
+      return ER_FAILED;
+    }
+  for (i = 0; i < num_class; i++)
+    {
+      class_names[i] = NULL;
+    }
+
+  for (i = 0; i < num_class; ++i)
+    {
+      if (fgets ((char *) buffer, DB_MAX_IDENTIFIER_LENGTH, input_file) == NULL)
+	{
+	  status = ER_FAILED;
+	  goto end;
+	}
+
+      ptr = strchr (buffer, '\n');
+      if (ptr)
+	{
+	  len = CAST_BUFLEN (ptr - buffer);
+	}
+      else
+	{
+	  len = (int) strlen (buffer);
+	}
+
+      if (len < 1)
+	{
+	  status = ER_FAILED;
+	  goto end;
+	}
+
+      class_names[i] = (char *) malloc (DB_SIZEOF (char) * (len + 1));
+      if (class_names[i] == NULL)
+	{
+	  status = ER_FAILED;
+	  goto end;
+	}
+
+      strncpy (class_names[i], buffer, len);
+      class_names[i][len] = 0;
+    }
+
+  status = get_class_mops (class_names, num_class, class_list, num_class_mops);
+
+end:
+
+  if (input_file)
+    {
+      fclose (input_file);
+    }
+
+  if (class_names)
+    {
+      for (i = 0; i < num_class; i++)
+	{
+	  if (class_names[i] != NULL)
+	    {
+	      free (class_names[i]);
+	      class_names[i] = NULL;
```

Comment:
Suggestion: use of **free_and_init ()  macro** for two statements
`free_and_init (class_names[i]);`